Add config linting script that checks for bool casing

anoadragon453 opened the pull request:
Add a linting script that enforces all boolean values in the default config be lowercase.

This has annoyed me for a while so I decided to fix it.

~~Pipelines repo PR: https://github.com/matrix-org/pipelines/pull/12~~ Not necessary anymore.

richvdh commented on scripts-dev/config-lint.sh:
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
```
it might be simpler just to grep for the incorrect formats, rather than try to correct them?

anoadragon453 replied:
I considered this at first, but I wanted to be able to add this to `scripts-dev/lint.sh` which could get run in a pre-commit hook for instance.

richvdh replied:
I'm not quite sure that follows... `lint.sh` doesn't correct for incorrect imports or pep8 failures, it just complains if they are wrong.

There might be value in having a script to correct the problems, but given how often it's likely to be a problem versus the added complexity of this script, I don't think it's worthwhile?

richvdh commented on scripts-dev/config-lint.sh:
```diff
@@ -0,0 +1,9 @@
+# Exits with 0 if there are no problems, or another code otherwise.
+
+# Fix non-lowercase true/false values
+sed -i -E "s/(#.*): +True(.*)/\1: true\2/g; s/(#.*): +False(.*)/\1: false\2/g;" synapse/config/*.py
```
1. why do we require a #
2. some of the groups are redundant:
```suggestion
sed -i -E "s/(#.*): +True/\1: true/g; s/(#.*): +False/\1: false/g;" synapse/config/*.py
```

anoadragon453 replied:
The code can contain false positives without the hash, such as docstrings:

https://github.com/matrix-org/synapse/blob/b617864cd9f81109e818bc5ae95bee317d917b72/synapse/config/logger.py#L199

richvdh replied:
ahhh. we should probably run it on the sample config rather than attempt to match only the config bits within `synapse/config`.

richvdh commented on tox.ini:
```diff
@@ -124,6 +124,9 @@ skip_install = True
 commands = /bin/sh -c "isort -c -df -sp setup.cfg -rc synapse tests"
 
+[testenv:check_config]
+commands = {toxinidir}/scripts-dev/lint-config.sh
```
Why not just add this to the `check_codestyle` env?

anoadragon453 replied:
I believe my thought process is that we'd have other config-related checks in the future, but you're right in that there's not a need to burn another agent when we have a codestyle one already.

anoadragon453 commented:
> I think it's worth noting that your script has missed
> 
> https://github.com/matrix-org/synapse/blob/0155478eb1ad6117d8339ad7b68b6d97bfe96e49/docs/sample_config.yaml#L414
> .

Yep, had to fix this case manually. I've also switched it so that the script just fixes the sample config instead of your code. I also changed the `git diff` to only check the sample config, so that it doesn't print out any changes you've made otherwise.
